extends Node

const BASE_COLOR = [
	Color(0.2, 0.2, 0.2, 1),
	Color(0.4, 0.4, 0.4, 1),
	Color(0.6, 0.6, 0.6, 1),
	Color(0.8, 0.8, 0.8, 1)
]

var RecolorShader := preload("res://Shader/recolor.gdshader")

var _color_table: Dictionary


func _ready():
	_load_color_table()


func create_recolor_material_from(color_name: String) -> ShaderMaterial:
	if color_name not in _color_table:
		printerr("color not in color table: %s" % color_name)
		return null

	var replace_colors = _get_color_platte_from(color_name)

	var material = ShaderMaterial.new()
	material.shader = RecolorShader

	material.set_shader_parameter("original_0", BASE_COLOR[0])
	material.set_shader_parameter("original_1", BASE_COLOR[1])
	material.set_shader_parameter("original_2", BASE_COLOR[2])
	material.set_shader_parameter("original_3", BASE_COLOR[3])

	material.set_shader_parameter("replace_0", replace_colors[0])
	material.set_shader_parameter("replace_1", replace_colors[1])
	material.set_shader_parameter("replace_2", replace_colors[2])
	material.set_shader_parameter("replace_3", replace_colors[3])

	return material


func _get_color_platte_from(color_name: String) -> PackedColorArray:
	if color_name not in _color_table:
		printerr("color not in color table: %s" % color_name)
		return PackedColorArray()
	return _color_table[color_name]


func _load_color_table() -> void:
	var table: ColorTable = preload("res://Systems/color_table.tres")

	for color_name in table.colors.keys():
		_color_table[color_name] = table.colors[color_name]


func _gradient_offset(size: int):
	var array = []
	for i in range(0, size):
		array.append(i / float(size))
	return array
